chore(bpfassets): Refactor to reduce the API exposed by bpfassets

dave-tucker opened the pull request:
Currently `pkg/bpfassets` contains many global variables that are accessed from many other packages.
This makes it hard for both the `bpfAssets` package and other packages that depend on it to be effectively unit tested.
To solve this, I've shrunk the API of be based on:

```go 
type Attacher interface {
	HardwareCountersEnabled() bool
	Detach()
	CollectProcesses() ([]ProcessBPFMetrics, error)
	CollectCPUFreq() (map[int32]uint64, error)
	GetEnabledBPFHWCounters() []string
	GetEnabledBPFSWCounters() []string
}
```

Where `NewAttacher()` will provide a new eBPF attacher and `NewMockAttacher()` is used for unit testing (dependency injection).

In un-picking access to globals it transpired that a `HwCountersEnabled()` was being accessed from many places in the codebase. It defaults to `true`. Once the eBPF probes have been loaded we mark it to `false` if we're unable to open perf events.

In order to keep this as consistent as possible, I've changed the program start ordering to load the probes first so the `HwCountersEnabled` bool can be propagated where it needs to go - and we can be sure it won't change after the fact.

Similarly, `GetEnabledBPFHWCounters()` and `GetEnabledBPFSWCounter()` were also changing after the probes were loaded, which is now also solved by loading eBPF earlier in the program startup.

The older BCC test has now been fixed to work with libbpf - and will hopefully run in CI.

The last commit in this series fixes test run control to allow you to:
`make VERBOSE=1 test` to run unit tests, bpf test + benchmarks, or to selectively run one of the 3 component parts.
e2e tests are excluded from that make target given they will fail without and e2e environment set up.

Requires: #1390
Updates: #1412

dave-tucker commented:
The test failure is due to the code that looks up the libbpf path. This has been an issue in my dev environment since day 1 😱 
I'll open a separate PR to fix that.

rootfs commented:
@dave-tucker this is great, can we have this soon so we can include it in the next release on Thu? Thanks

dave-tucker commented on Makefile:
```diff
@@ -229,17 +231,20 @@ cross-build: clean_build_local cross-build-linux-amd64 cross-build-linux-arm64 c
 	mkdir -p $(GOBIN)
 	mkdir -p $(ENVTEST_ASSETS_DIR)
 	@test -f $(ENVTEST_ASSETS_DIR)/ginkgo || \
 	 (go install -mod=mod github.com/onsi/ginkgo/v2/ginkgo@v2.4.0  && \
```
tracked in https://github.com/sustainable-computing-io/kepler/issues/1428

dave-tucker commented on pkg/bpfassets/attacher/attacher.go:
```diff
@@ -192,141 +159,176 @@ func attachLibbpfModule() (*bpf.Module, error) {
+	for _, arrayName := range toResize {
+		err = resizeArrayEntries(a.module, arrayName, a.cpuCores)
 		if err != nil {
 			klog.Infof("failed to resize array %s: %v\n", arrayName, err)
```
We didn't error out previously, however I do think it would make sense to do so. I'll fix that.

dave-tucker commented on pkg/bpfassets/attacher/attacher.go:
```diff
@@ -192,141 +159,176 @@ func attachLibbpfModule() (*bpf.Module, error) {
-		return libbpfModule, fmt.Errorf("failed to get kepler_read_page_trace: %v", err)
+		return fmt.Errorf("failed to get kepler_read_page_trace: %v", err)
 	} else {
 		if _, err = page_read.AttachGeneric(); err != nil {
```
> Query: why some places used Attachgeneric and other places used AttachTracepoint ?

No reason. Fixed to use `AttachGeneric`.

> in case kepler_write_page_trace fails to attach, and kepler_read_page_trace succeeds there will be no entry for config.PageCacheHit in enabledSoftwareCounters

That's correct. I'm not sure that it makes sense to have only one of these 2 programs attached.
They both increase the `PageCacheHit` counter... but having only one is going to skew the results.

dave-tucker commented on pkg/bpfassets/attacher/types.go:
```diff
@@ -20,6 +20,17 @@ import (
 
+var SoftIRQEvents = []string{config.IRQNetTXLabel, config.IRQNetRXLabel, config.IRQBlockLabel}
+
+type Attacher interface {
```
I agree. I'll do this in a follow up as the diff is already pretty big.

dave-tucker commented:
Thanks for the review @vimalk78. I've addressed your comments. The only outstanding thing left to clarify is what the "correct" behaviour should be in the case that either of the probes loaded for tracking `PageCacheHit` cannot be loaded.

vimalk78 commented:
>  in the case that either of the probes loaded for tracking `PageCacheHit` cannot be loaded.

since we return error in case `kepler_sched_switch_trace` fails to load, how about returning error in case `kepler_write_page_trace`  or `kepler_read_page_trace` ?

sthaha commented:
@dave-tucker Thanks for the PR. It looks good to me. I ran the version against the `upstream/main` using the `dev` docker-compose today and got "interesting" results. The variance between 2 versions of kepler running (kepler:latest and this PR) is quite a bit. 

We may want to investigate further why this is the case without blocking this PR. It may very well be pointing to some other underlying issue in kepler / bpf  or even the (docker compose) setup.  

<img width="1909" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/3005132/0e0359f1-8fd5-4faf-bf95-d76c513e1e0e">

<img width="1919" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/3005132/ac9f8dc6-825d-429e-bf71-9c3126793a08">


I have created a PR for the "enhanced" dashboard -  https://github.com/sustainable-computing-io/kepler/pull/1429
